### ISSUE TITLE
All (*) ability allows access to abilities in a different scope

### ISSUE THREAD

**Issue by pyr0hu** [JosephSilber/bouncer#554]
My current use case is a multi-tenant application and I have several developer/admin permission that manages access to debug stuff (e.g. horizon, telescope). I allow the `all abilities` * to the tenant admin (which is different from the system admin) to manage all their stuff that belongs to the tenant. 

Tenancy is solved using scopes.

But because of the *, Bouncer allows access to the developer abilities but they are not in the same scope. The developer abilities are scoped to `null` while the tenant ones are scoped to `tenant_id`.

Is there any way to solve this? I could just drop `->isA()` calls everywhere, but I'd rather solve this using `->can()`s

**Comment by JosephSilber** [comment]
It's unclear to me whether you're reporting a bug or requesting a feature.

> All (*) ability allows access to abilities in a different scope

Are you saying that's how it currently works, or how you want it to work?

**Comment by pyr0hu** [comment]
Sorry, my bad if it wasn't clear. Let me rephrase this.

I ain't sure if this is a bug or a feature, but I'm leaning towards it's a question actually.

What I wanted to say, scoped abilities has a higher "priority" then the unscoped ones. If I allow an ability for a user with a `null` scope then forbid the same ability for a role with scope 1, then set the scope to null and check if the user `->can()`, it returns false.

Concrete example:

I have two roles: a `developer` which has a scope of `null` and has access to an ability called `access-horizon`.
I also have a `tenant-admin` role with a scope of `tenant-1`. The `tenant-admin` role has an `*` ability. This is also scoped, so in the permissions table, the `all` ability has a scope of `tenant-1`.

My user, let's call it `Gandalf`, has the roles of `developer`. Now when he wants to access Horizon, he can view it. All good. Now consider `Frodo`. He has the role of `tenant-admin`. He opens up Horizon and still has access. `Gandalf` does not want to give access to `Frodo` to Horizon, so I forbid the `tenant-admin` role the ability `access-horizon`. Now `Frodo` cannot access it, all good. Here comes `Legolas` who got the roles `developer` AND `tenant-admin`. He tries to access Horizon, but receives Forbidden.

I tried to modify my Horizon gate to set the scope to null when checking the ability, but it still returned false and did not let `Legolas` to view Horizon. I'm using the simplest `Bouncer::scope()` without any of the modifiers

**Comment by JosephSilber** [comment]
A scope of `null` means "all scopes".

---

I would advise you to instead create a separate `landlord` scope value:

```php
Bouncer::scope()->onceTo('landlord', function () {
    Bouncer::allow('developer')->to('access-horizon');
});
```

And use that in your Horizon check:


```php
return Bouncer::scope()->onceTo('landlord', fn () => Gate::can('access-horizon'));
```
